### PR TITLE
Updated RigidBodyMotion Objects

### LINKED ISSUE #7233: Drop ComputeGrainCenterUserObject Refactor Volume and Centers Calcs
### Description of the enhancement or error report

The ComputeGrainCenterUserObject in the phase_field module doesn't work when n_grains > n_ops and overlaps in functionality "mostly" provided by the GrainTracker already.  Implement a better interface for centers and volumes in the GrainTracker and remove this object.

### Rationale for the enhancement or information for reproducing the error

Object only works for a limited set of small test problems.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

None.

Tagging @SudiptaBiswas and @itgreenquist

### PR DESCRIPTION
**Pull request by SudiptaBiswas**
- Modifies all the rigid body motion objects in PhaseField module to incorporate correct jacobian calculation as per #7216
- Maps RBM objects to `grain_num` instead of `op_num` using GrainTracker
- Closes #7233 & #5762

### CODE REVIEW

**Comment by jwpeterson** [comment]
Looks like this needs to be rebased after the merge in 8635e2c?

**Comment by SudiptaBiswas** [comment]
Yeah, just updated it. It's ready for review now.

File: modules/phase_field/tests/rigidbodymotion/tests [hunk 1]
```diff
@@ -1,49 +1,51 @@
...
     input = 'grain_forcedensity.i'
     exodiff = 'grain_forcedensity_out.e'
+    allow_warnings = true
+  [../]
+
+  [./grain_motion2]
+    type = 'Exodiff'
+    input = 'grain_motion2.i'
+    exodiff = 'grain_motion2_out.e'
+    allow_warnings = true
```
Comment (permcody) [inline modules/phase_field/tests/rigidbodymotion/tests]: Just curious, what warnings are you getting on these tests?

**Comment by permcody** [comment]
After this is merged, we should be able fixup and merge https://github.com/idaholab/moose/issues/7233 too, correct?

File: modules/phase_field/tests/rigidbodymotion/tests (hunk 1 above)
Comment (SudiptaBiswas) [inline modules/phase_field/tests/rigidbodymotion/tests]: We have warnings in `NonlocalKernel `and `ShapeElementUserObject` about the nonlocal jacobian being computationally expensive. That would reflect in these tests as well.

**Comment by SudiptaBiswas** [comment]
Yes, this takes care of #7233. Instead of dropping I have deprecated `ComputeGrainCenterUserObject` though.

**Comment by permcody** [comment]
@SudiptaBiswas - @dschwen just merged my huge GrainTracker update - AGAIN!

I hope that it doesn't foul you up on this PR. Maybe before this is merged you can run the tests locally again and verify everything is still OK? If you have any questions, let me know.

**Comment by SudiptaBiswas** [comment]
I will rebase by branch and see how the tests go.

**Comment by SudiptaBiswas** [comment]
All the tests are fine in serial. However, some of the tests does not even converge in parallel. Probably, kernels are not getting correct values from GrainTracker in parallel. I will dig in further to pin point the issue.

**Comment by permcody** [comment]
Handy tip: 
If you need to see what's happening on the remaining ranks try using the options:
--keep-cout --redirect-stdout. You'll get individual output files for each rank. You can just output what you need and compare

**Comment by SudiptaBiswas** [comment]
The grain volumes & the grain centers are not same accross processors. Here is an example how they look,
```
processor_id    grain_id    volume     centroid (x,y)
    0              0         375       (20, 6.58333)
    1              0         112.5     (527.5, 107.5)
    0              1         375       (30, 18.4167)
    1              1         262.5     (1427.5, 762.5)
```
The values on rank 0 is correct, but it's not communicated to other ranks correctly.

**Comment by permcody** [comment]
Oh! you need volume and centroid on all processors? I definitely didn't account for that. It can be easily fixed. Sorry about that. Am I missing anything else?

**Comment by SudiptaBiswas** [comment]
>
> Oh! you need volume and centroid on all processors?
>
Yes, I do. I have created an issue for this.


> Am I missing anything else?
>
 Will be able to figure out after I have the volume and centroid correct.

**Comment by permcody** [comment]
Working on it now

**Comment by permcody** [comment]
I'm glad this passed! Unfortunately my PR didn't due to a silly bug related to the EBSD logic that I just fixed. I just rebased and pushed new commits but I actually think your branch will rebase OK on devel once mine gets merged because I didn't change these two commits so they should resolve when git does the rebase. We'll see.

**Comment by SudiptaBiswas** [comment]
This is done from my side and up for review.

**Comment by permcody** [comment]
Excellent

File: modules/phase_field/include/kernels/GrainRigidBodyMotionBase.h [hunk 2]
```diff
@@ -7,56 +7,82 @@
...
+  const GrainForceAndTorqueInterface & _grain_force_torque;
+  const std::vector<RealGradient> & _grain_forces;
+  const std::vector<RealGradient> & _grain_torques;
+  const std::vector<Real> & _grain_force_c_jacobians;
+  const std::vector<std::vector<Real> > & _grain_force_eta_jacobians;
+
+  /// constant value corresponding to grain translation
+  Real _mt;
+  /// constant value corresponding to grain rotation
+  Real _mr;
```
Comment (dschwen) [inline modules/phase_field/include/kernels/GrainRigidBodyMotionBase.h]: If it is constant make them `const`

File: modules/phase_field/src/userobjects/ComputeGrainForceAndTorque.C [hunk 3]
```diff
@@ -4,100 +4,153 @@
...
+          if (j == _grain_tracker.getOpToGrainsVector(_current_elem->id())[k])
+          {
+            const auto centroid = _grain_tracker.getGrainCentroid(j);
+            for (_qp=0; _qp<_qrule->n_points(); ++_qp)
+              if ((*_dFdgradeta[i])[_qp][j] != 0.0)
+              {
+                const RealGradient compute_torque_jacobian_eta =_JxW[_qp] * _coord[_qp] * (_q_point[_qp] - centroid).cross(_grad_phi[_j][_qp] * (*_dFdgradeta[i])[_qp][k]);
+                _force_torque_eta_jacobian_store[i][(6*j+0)*_total_dofs+_j_global] += _JxW[_qp] * _coord[_qp] * _grad_phi[_j][_qp](0) * (*_dFdgradeta[i])[_qp][k];
+                _force_torque_eta_jacobian_store[i][(6*j+1)*_total_dofs+_j_global] += _JxW[_qp] * _coord[_qp] * _grad_phi[_j][_qp](1) * (*_dFdgradeta[i])[_qp][k];
+                _force_torque_eta_jacobian_store[i][(6*j+2)*_total_dofs+_j_global] += _JxW[_qp] * _coord[_qp] * _grad_phi[_j][_qp](2) * (*_dFdgradeta[i])[_qp][k];
```
Comment (dschwen) [inline modules/phase_field/src/userobjects/ComputeGrainForceAndTorque.C]: do 
```
const Real factor = _JxW[_qp] * _coord[_qp] * (*_dFdgradeta[i])[_qp][k]);
```
and use that in all 4 terms.

File: modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C [hunk 4]
```diff
@@ -22,50 +23,191 @@ MultiGrainRigidBodyMotion::MultiGrainRigidBodyMotion(const InputParameters & par
...
+    unsigned int grain_index = grain_indices[i];
+    if (grain_index != libMesh::invalid_uint)
+    {
+      const auto volume = _grain_tracker.getGrainVolume(grain_index);
+      const auto centroid = _grain_tracker.getGrainCentroid(grain_index);
+
+      velocity_advection += _mt / volume * ((*_vals[i])[_qp] * _grain_forces[grain_index])
+                            + _mr / volume * (_grain_torques[grain_index].cross(_q_point[_qp] - centroid)) * (*_vals[i])[_qp];
+      div_velocity_advection += _mt / volume * ((*_grad_vals[i])[_qp] * _grain_forces[grain_index])
+                                + _mr / volume * (_grain_torques[grain_index].cross(_q_point[_qp] - centroid)) * (*_grad_vals[i])[_qp];
```
Comment (dschwen) [inline modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C]: precompute `_mr / volume * (_grain_torques[grain_index].cross(_q_point[_qp] - centroid))`

File: modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C [hunk 5]
```diff
@@ -22,50 +23,191 @@ MultiGrainRigidBodyMotion::MultiGrainRigidBodyMotion(const InputParameters & par
...
+    if (grain_index != libMesh::invalid_uint)
+    {
+      const auto volume = _grain_tracker.getGrainVolume(grain_index);
+      const auto centroid = _grain_tracker.getGrainCentroid(grain_index);
+      getUserObjectCJacobians(jdof, grain_index);
+
+      velocity_advection += _mt / volume * _grain_forces[grain_index] * (*_vals[i])[_qp]
+                            + _mr / volume * _grain_torques[grain_index].cross(_q_point[_qp] - centroid) * (*_vals[i])[_qp];
+      div_velocity_advection += _mt / volume * _grain_forces[grain_index] * (*_grad_vals[i])[_qp]
+                                + _mr / volume * _grain_torques[grain_index].cross(_q_point[_qp] - centroid) * (*_grad_vals[i])[_qp];
```
Comment (dschwen) [inline modules/phase_field/src/kernels/MultiGrainRigidBodyMotion.C]: same here

**Comment by permcody** [comment]
👍  I will merge this once the build is complete.

**Pull request merged**